### PR TITLE
Make unique resource types to enable Primary Key Type Behaviour.

### PR DESCRIPTION
**Pull request by fppt**
- Allow creating unique resource types whose instances can only have ONE relation. 
- Enforce this rule when creating relations.
- Various clean ups of old Core API code.

### CODE REVIEW

**Comment by aelred** [comment]
Perhaps we should also have another (I know, sorry) interface for `UniqueResourceType` and `UniqueResource`? Mainly because otherwise there is no API benefit in this change.

Currently, we do this to look up an entity by a resource:
```
ResourceType<String> title = graph.getResourceType("title", DataType.STRING);
Instance apocalypseNow = graph.getResource(title, "Apocalypse Now").ownerInstances().iterator().next();
```

I'd rather do this:
```
UniqueResourceType title = graph.getUniqueResourceType("title", DataType.STRING);
Instance apocalypseNow = graph.getUniqueResource(title, "Apocalypse Now").owner();
```

**Comment by fppt** [comment]
Not a bad idea to be honest. Let's get a requirement going and see if everyone is happy with it before adding another object to our model. For now are you okay with putting this one in ?

File: mindmaps-core/src/main/java/io/mindmaps/concept/Rule.java [hunk 1]
```diff
@@ -58,14 +58,14 @@
      *
      * @return
      */
-    boolean getExpectation();
+    Boolean getExpectation();
```
Comment (aelred) [inline mindmaps-core/src/main/java/io/mindmaps/concept/Rule.java]: ok

File: mindmaps-graph/src/main/java/io/mindmaps/graph/internal/ConceptImpl.java [hunk 2]
```diff
@@ -556,8 +543,18 @@ T setProperty(Schema.ConceptProperty key, Object value){
      * @param key The key of the non-unique property to retrieve
      * @return The value stored in the property
      */
-    public Object getProperty(Schema.ConceptProperty key){
-        return getProperty(key.name());
+    @SuppressWarnings("unchecked")
+    public <X extends Object> X getProperty(Schema.ConceptProperty key){
```
Comment (aelred) [inline mindmaps-graph/src/main/java/io/mindmaps/graph/internal/ConceptImpl.java]: let the generic wizardry begin!

**Pull request merged**